rustls/src/crypto/aws_lc_rs/ticketer.rs

```diff
@@ -0,0 +1,354 @@
+            6 * 60 * 60,
+            make_ticket_generator,
+        )?))
+    }
+
+    /// Make the recommended `Ticketer`.  This produces tickets
+    /// with a 12 hour life and randomly generated keys.
+    ///
+    /// The `Ticketer` uses the [RFC 5077 §4] "Recommended Ticket Construction",
+    /// using AES 128 for encryption and HMAC-SHA256 for ciphertext authentication.
+    ///
+    /// [RFC 5077 §4]: https://www.rfc-editor.org/rfc/rfc5077#section-4
+    #[cfg(not(feature = "std"))]
+    pub fn new<M: crate::lock::MakeMutex>(
+        time_provider: &'static dyn TimeProvider,
+    ) -> Result<Arc<dyn ProducesTickets>, Error> {
+        Ok(Arc::new(crate::ticketer::TicketSwitcher::new::<M>(
+            6 * 60 * 60,
+            make_ticket_generator,
+            time_provider,
+        )?))
+    }
+}
+
+fn make_ticket_generator() -> Result<Box<dyn ProducesTickets>, GetRandomFailed> {
+    // NOTE(XXX): Unconditionally mapping errors to `GetRandomFailed` here is slightly
+    //   misleading in some cases (e.g. failure to construct a padded block cipher encrypting key).
+    //   However, we can't change the return type expected from a `TicketSwitcher` `generator`
+    //   without breaking semver.
+    Ok(Box::new(Rfc5077::new().map_err(|_| GetRandomFailed)?))
+}
+
+/// An RFC 5077 "Recommended Ticket Construction" implementation of a [`Ticketer`].
+struct Rfc5077 {
+    aes_key: [u8; cipher::AES_128_KEY_LEN],
+    hmac_key: hmac::Key,
+    key_name: [u8; 16],
+    lifetime: u32,
+    maximum_ciphertext_len: AtomicUsize,
+}
+
+impl Rfc5077 {
+    fn new() -> Result<Self, Error> {
+        let rand = SystemRandom::new();
+
+        // Generate a random AES 128 key to use for AES CBC encryption.
+        let mut aes_key = [0u8; cipher::AES_128_KEY_LEN];
+        rand.fill(&mut aes_key)
+            .map_err(|_| GetRandomFailed)?;
+
+        // Generate a random HMAC SHA256 key to use for HMAC authentication.
+        let hmac_key = hmac::Key::generate(hmac::HMAC_SHA256, &rand).map_err(unspecified_err)?;
+
+        // Generate a random key name.
+        let mut key_name = [0u8; 16];
+        rand.fill(&mut key_name)
+            .map_err(|_| GetRandomFailed)?;
+
+        Ok(Self {
+            aes_key,
+            hmac_key,
+            key_name,
+            lifetime: 60 * 60 * 12,
+            maximum_ciphertext_len: AtomicUsize::new(0),
+        })
+    }
+}
+
+impl ProducesTickets for Rfc5077 {
+    fn enabled(&self) -> bool {
+        true
+    }
+
+    fn lifetime(&self) -> u32 {
+        self.lifetime
+    }
+
+    /// Encrypt `message` and return the ciphertext.
+    fn encrypt(&self, message: &[u8]) -> Option<Vec<u8>> {
+        // Convert the raw AES 128 key bytes into an encrypting key using CBC PKCS#7 padding.
+        let aes_key = cipher::UnboundCipherKey::new(&cipher::AES_128, &self.aes_key[..]).ok()?;
+        let aes_key = cipher::PaddedBlockEncryptingKey::cbc_pkcs7(aes_key).ok()?;
```

Comment:
Can we keep this instead of `self.aes_key` to avoid paying the key schedule costs for each encryption/decryption? (probably means keeping both a `PaddingBlockEncryptingKey` and a `PaddedBlockDecryptingKey`)

Reply:
Done.